src/main/kotlin/cacophony/grammars/Analysis.kt

```diff
@@ -1,44 +1,169 @@
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-): Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>> {
-    TODO("Not implemented")
+sealed class Entity<out Q, out A, out R> {
+    data class Ref<Q, A, R>(val ref: DFAStateReference<Q, A, R>) : Entity<Q, A, R>()
+
+    data class Sym<A>(val sym: A) : Entity<Nothing, A, Nothing>()
 }
 
-fun <StateType, SymbolType> findFirst(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-    nullable: Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>>,
-): Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>> {
+typealias StateToSymbolsMap<StateType, SymbolType, ResultType> =
+    Map<DFAStateReference<StateType, SymbolType, ResultType>, Collection<SymbolType>>
+
+fun <StateType, SymbolType, ResultType> findNullable(
+    automata: Map<SymbolType, DFA<StateType, SymbolType, ResultType>>,
+): Collection<DFAStateReference<StateType, SymbolType, ResultType>> {
     TODO("Not implemented")
 }
 
-fun <StateType, SymbolType> findFollow(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-    nullable: Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>>,
-    first: Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>>,
-): Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>> {
+class FindFirstImpl<StateType, SymbolType, ResultType>(
+    automata: Map<SymbolType, DFA<StateType, SymbolType, ResultType>>,
+    nullable: Collection<DFAStateReference<StateType, SymbolType, ResultType>>,
+) {
+    private val stateReferenceToSymbol = automata.map { (symbol, dfa) -> Pair(Pair(dfa.getStartingState(), dfa), symbol) }.toMap()
+    private val symbolToStateReference = stateReferenceToSymbol.map { (u, v) -> Pair(v, u) }.toMap()
+    val firstMap: Map<DFAStateReference<StateType, SymbolType, ResultType>, Set<SymbolType>>
+
+    private fun fromState(ref: DFAStateReference<StateType, SymbolType, ResultType>): Entity<StateType, SymbolType, ResultType> {
+        val sym: SymbolType? = stateReferenceToSymbol[ref]
+        return if (sym == null) {
+            Entity.Ref(ref)
+        } else {
+            Entity.Sym(sym)
+        }
+    }
+
+    private fun fromSymbol(sym: SymbolType): Entity<StateType, SymbolType, ResultType> {
+        return Entity.Sym(sym)
+    }
+
+    private fun toState(entity: Entity<StateType, SymbolType, ResultType>): DFAStateReference<StateType, SymbolType, ResultType>? {
+        if (entity is Entity.Ref) {
+            return entity.ref
+        }
+        if (entity is Entity.Sym) {
+            return symbolToStateReference[entity.sym]
+        }
+        return null
+    }
+
+    private fun toSymbol(entity: Entity<StateType, SymbolType, ResultType>): SymbolType? {
```

Comment:
`fromState`, `fromSymbol`, `toState`, `toSymbol` should be methods of `Entity` class

Reply:
fromSymbol and fromState are not member of Entity (or rather GeneralizedSymbol as it is now called) to allow them for implicit access to stateReferenceToSymbol and symbolToStateReference maps